### PR TITLE
Support lists of models

### LINKED ISSUE #240: `list[MyPydanticModel]` does not appear to go through downgrades
**Describe the bug**

It appears that if a route returns a list of models (and not a pydantic model directly), these do not go through the standard response transformations.

Syntax to trigger bug:

```python
def do_something() -> list[MyPydanticModel]:
    ...
```

**To Reproduce**

## `bug.py`

<details>
<summary>Show code</summary>

```python
import typing

from pydantic import BaseModel
from cadwyn import Cadwyn, Version, VersionBundle, VersionedAPIRouter, VersionChange, \
    ResponseInfo, convert_response_to_previous_version_for, schema


class Thing(BaseModel):
    thing: bool


class UseBools(VersionChange):
    description = "The previous API version used the words 'yes' and 'no'."
    instructions_to_migrate_to_previous_version = (
        schema(Thing).field("thing").had(type=typing.Literal["yes", "no"]),
    )

    @convert_response_to_previous_version_for(Thing)
    def downgrade_response(response: ResponseInfo) -> None:
        response.body["thing"] = "yes" if response.body["thing"] else "no"


router = VersionedAPIRouter()

@router.get("/thing")
def get_thing() -> Thing:
    """Control: Demonstrates that downgrades work in general."""
    return {"thing": True}


# ----------------------------------------------------------------------------------
# This route demonstrates the bug being reported.
#
@router.get("/things")
def get_things() -> list[Thing]:
    """Demonstrate that the type annotation `list[Thing]` does not get converted."""
    return [
        {"thing": True},
    ]
# ----------------------------------------------------------------------------------


app = Cadwyn(
    versions=VersionBundle(
        Version("2025-02-02", UseBools),
        Version("2025-01-01"),
    )
)
app.generate_and_include_versioned_routers(router)
```

</details>

## Commands to run

Terminal 1:

```
fastapi dev bug.py
```

Terminal 2:

```console
# New version -- returns bools
$ curl -s --header 'Content-Type: application/json' --header 'X-API-VERSION: 2025-02-02' http://127.0.0.1:8000/thing
{"thing":true}

$ curl -s --header 'Content-Type: application/json' --header 'X-API-VERSION: 2025-02-02' http://127.0.0.1:8000/things
[{"thing":true}]

# Old version -- uses "yes" and "no"
$ curl -s --header 'Content-Type: application/json' --header 'X-API-VERSION: 2025-01-01' http://127.0.0.1:8000/thing
{"thing":"yes"}

$ curl -s --header 'Content-Type: application/json' --header 'X-API-VERSION: 2025-01-01' http://127.0.0.1:8000/things
Internal Server Error
```

## Output from terminal 1

```
fastapi.exceptions.ResponseValidationError: 1 validation errors:
  {'type': 'literal_error', 'loc': ('response', 0, 'thing'), 'msg': "Input should be 'yes' or 'no'", 'input': True, 'ctx': {'expected': "'yes' or 'no'"}}

```

**Expected behavior**
I expected that the type annotation supported by FastAPI would work.

**Operating system**

<details>
<summary>System details</summary>

```console
$ lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 22.04.5 LTS
Release:        22.04
Codename:       jammy

$ python -V
Python 3.13.1

$ pip list
Package             Version    Editable project location
------------------- ---------- ----------------------------------
annotated-types     0.7.0
anyio               4.7.0
cadwyn              4.4.3
certifi             2024.8.30
click               8.1.7
demo-api-versioning 20241209.1 /home/kurt/dev/demo-api-versioning
dnspython           2.7.0
email_validator     2.2.0
fastapi             0.115.6
fastapi-cli         0.0.6
h11                 0.14.0
httpcore            1.0.7
httptools           0.6.4
httpx               0.28.1
idna                3.10
issubclass          0.1.2
Jinja2              3.1.4
markdown-it-py      3.0.0
MarkupSafe          3.0.2
mdurl               0.1.2
orjson              3.10.12
pip                 24.3.1
pydantic            2.10.3
pydantic_core       2.27.1
Pygments            2.18.0
python-dotenv       1.0.1
python-multipart    0.0.19
PyYAML              6.0.2
rich                13.9.4
rich-toolkit        0.12.0
setuptools          75.6.0
shellingham         1.5.4
sniffio             1.3.1
starlette           0.41.3
typer               0.15.1
typing_extensions   4.12.2
uvicorn             0.32.1
uvloop              0.21.0
watchfiles          1.0.0
websockets          14.1
wheel               0.45.1
```

</details>

### PR DESCRIPTION
**Pull request by kurtmckee**
This PR allows routes to be annotated with:

```python
@router.get("/things")
def get_things() -> list[Thing]:
    ...
```

Fixes #240

### CODE REVIEW

**Comment by kurtmckee** [comment]
As noted in #240, this may not be an ideal solution; for example, each iterated call to `self._migrate_response()` creates a new `ResponseInfo` instance, which is immediately discarded because only its `.body` is needed.

It's possible that the conversion work could be done by a helper method, so that it's not `self._migrate_response()` that's getting called.

Please let me know if this is even desirable work, and if you'd like to see an improvement in the design here. :+1:

File: tests/test_data_migrations.py [hunk 1]
```diff
@@ -1235,3 +1236,38 @@ def response_converter(response: ResponseInfo):
...
+
+    # `get_thing()`, singular, is not critical to this test.
+    # It exists to avoid `RouteResponseBySchemaConverterDoesNotApplyToAnythingError`.
+    @router.get("/thing")
+    async def get_thing() -> Thing:  # pragma: no cover
+        return {"thing": True}  # pyright: ignore[reportReturnType]
+
+    @router.get("/things")
+    async def get_things() -> list[Thing]:
+        return [{"thing": True}]  # pyright: ignore[reportReturnType]
```
Comment (zmievsa) [inline tests/test_data_migrations.py:1256]:
```suggestion
    @router.get("/things", response_model=list[Thing])
    async def get_things()
        return [{"thing": True}]
```

**Review by zmievsa — CHANGES_REQUESTED** [review]
Nice work! Just a few small comments from my side

File: cadwyn/structure/versions.py [hunk 2]
```diff
@@ -392,10 +393,27 @@ def _migrate_response(
         self,
         response_info: ResponseInfo,
         current_version: VersionDate,
-        head_response_model: type[BaseModel],
+        head_response_model: type[BaseModel] | type[list[BaseModel]],
         path: str,
         method: str,
     ) -> ResponseInfo:
+        # Handle `list[BaseModel]`.
+        if isinstance(head_response_model, types.GenericAlias) and isinstance(response_info.body, Sequence):
```
[the remark this answers is not in the document]
Comment (kurtmckee) [reply cadwyn/structure/versions.py:401]: I couldn't either, but I saw that `isinstance(..., Sequence)` checks were already in use elsewhere in the file, so I thought that might be the desired formulation.

I'll switch this to `list`.

**Comment by kurtmckee** [comment]
I want to test one or two additional items before this merges (if it's a desirable change); I'm particularly interested in this use case:

```python
class Thing(BaseModel):
    ...


class TopLevel(BaseModel):
    things: list[Thing]  # <-- Will this get converted?
```

I anticipate that returning a JSON array (instead of an object) may be a niche use case, but I bumped into that while preparing a rough-cut tech demo to present in my workplace, which is how I got to this PR.

**Comment by kurtmckee** [comment]
Hmm, this doesn't address the more common case that I described above. I think that responding with JSON arrays may be too niche a use case to incorporate into cadwyn.

If you think this PR is valuable, you're welcome to merge it. However, I recommend closing.

I'll work to submit a PR that allows nested objects to be converted, which is far and away the more common case that I expect to use at work. However, I'll open an issue first to document the target work.

**Pull request closed without merging**